### ISSUE TITLE
Please stop using post-install messages as ads

### ISSUE THREAD

**Issue by olivierlacan** [harrowio/capistrano-harrow#4]
It's really distasteful. 

![image](https://cloud.githubusercontent.com/assets/65950/19471258/b49d8b8c-9522-11e6-9f5c-c5c66099e21d.png)

**Comment by leehambley** [comment]
Feel free to stop using Capistrano and use software the better meets your moralistic standards.

**Comment by olivierlacan** [comment]
I don't think it's "moralistic" to expect libraries not to abuse post-installation hooks meant to help installation of the library.

I have nothing against Capistrano.

**Comment by leehambley** [comment]
It took you more time to complain about this than it would have done to simply ignore it. 

We have to support our work somehow, and offering a hosted service (with an extremely generous free plan) which directly improves the use of Capistrano, whilst earning us a little money to invest most of our time into open source is the only feasible way we can do that without submitting to corporate sponsorship from a third party.

**Comment by mattbrictson** [comment]
@olivierlacan FYI, you can easily silence post-install messages per gem in Bundler 1.10+: http://bundler.io/blog/2015/06/24/version-1-10-released.html#new-features

**Comment by dhh** [comment]
😢

**Comment by leehambley** [comment]
Something constructive to add?

Am 29.01.2017 23:02 schrieb "David Heinemeier Hansson" <
notifications@github.com>:

**Comment by dhh** [comment]
What I've added is my sadness that the install hook has been turned into this. In my opinion, it's crass. And as someone who paid Jamis to develop Capistrano for years, it's sad to his and our work used in this way.

I humbly suggest that this won't achieve the effect that you're hoping from. I don't think this is going to drum up new business as much as its going to piss people off with its in-your-faceness and abuse of the install message.

Could you imagine if installing Rails came with a big fucking gem install hook ad for Basecamp? Crass too that would be.

But, it'd still like to help you guys promote Harrow.io. We have more than 100K people following rails on Twitter. Could I trade you the removal of this advertisement for a shout-out to that highly targeted and relevant group?

It's wonderful that you guys are maintaining Capistrano and I certainly welcome doing so off the back of a commercial business.

**Comment by dhh** [comment]
Additionally, making capistrano-harrow a direct and forced dependency of capistrano is equally over-the-top. And again, I don't think it'll get the desired effect. And again, imagine that installing rails can with a hard dependency on basecamp-api. I don't think anyone would find that in good taste either.

I think it's counterproductive to attempt to force people to use something rather than to entice them to do so. Harrow looks like a lovely product. It's got the authority of capistrano maintainers behind it. There are better ways to draw that connection than by attempting to shove it down people's throat.

I'd be happy to help brainstorm better ways.

**Comment by dhh** [comment]
I hope we can find some better ways. Otherwise I don't see how Rails can continue to promote Capistrano in the default Gemfile or documentation. That would be a shame.

**Comment by leehambley** [comment]
> What I've added is my sadness that the install hook has been turned into this. In my opinion, it's crass. And as someone who paid Jamis to develop Capistrano for years, it's sad to his and our work used in this way.

Understood. Although I beg you to remember that Jamis hasn't been involved with Capistrano since ~2008/9. I would say in business terms your investment paid off, and set the community going in the right direction, non the less I'll acknowledge that emotions don't always have to make sense, and I understand your position.

> I humbly suggest that this won't achieve the effect that you're hoping from. I don't think this is going to drum up new business as much as its going to piss people off with its in-your-faceness and abuse of the install message.

It converts really, really well, and we get a great deal of positive feedback (negative feedback is restricted to this issue and a couple of angry tweets when we first launched the integration), we've been thanked for innovative use of post install messages more than we've been berated for abusing them; again a concession from my side - I was somewhat pushed into trying this process through pushy investors who were tech savvy enough to know that this feature of Rubygems existed, and wanted to know why we weren't "exploiting it", the current implementation treads a line between being invasive and informative whilst still polite and easy to ignore for those who aren't interested. In my opinion. (one idea I had to shoot down was to put a prompt in every Capistrano invocation to ask people why they were still using it in the terminal rather than "in the cloud" 🥇 )

> Could you imagine if installing Rails came with a big fucking gem install hook ad for Basecamp? Crass too that would be.

Agreed, if I'm about to start a Rails project however there's a very, very weak argument to be made that I might be interested in trying a new project management approach for my apparent new project. 

The link between Capistrano and Harrow however is much, much stronger. You've not been involved for years supporting new Capistrano users in IRC, Google Groups StackOverflow, but new users who don't have the benefit of our years of experience struggle, **a lot** with Capistrano. They have the wind knocked right out of their sails. High on achievement, and the excitement of putting a their new creation live on the internet, they are thrown into a world of things they barely understand (let's name it "Unix Land"). I'm not sure what role you personally play in supporting newcomers to Rails, but a consistent problem in supporting the Capistrano community (especially the newcomers) is working out environmental issues, permissions, agent forwarding and security issues (namely SSH, keys, etc) Harrow was designed principally to be my way to support them, think of it as a kind of codepen/jsbin/etc for Capistrano. All Capistrano's features are 1:1 represented, environments tasks and dependencies, it's for lack of a better description "a web based Capistrano", of course since these things are expensive to build, it has a paid plan, and a small number of users support the platform for everyone else.

> But, it'd still like to help you guys promote Harrow.io. We have more than 100K people following rails on Twitter. Could I trade you the removal of this advertisement for a shout-out to that highly targeted and relevant group?

Thats an incredibly kind offer, one I'd like to take you up on very soon. We've been very fortunate with Harrow's growth so far, and we're actually intending to open-source a large (and functional, basically everything barring the billing subsystem) part of it very, very soon. It's been a difficult (and expensive!) birth of the project, and as we've found our niche, we're finally in a position where we can open source the platform at large without sabotaging the revenue streams which we rely on to keep building it. When we do this, I'd be very glad of your assistance with the announcement. It's a success story we'd all love to share with the community at large.

> It's wonderful that you guys are maintaining Capistrano and I certainly welcome doing so off the back of a commercial business.

That's good to hear. We joke around the office that you guys are a publishing business with a web app on the side! Joking aside atypical paths to commercial success have been a trial by fire for us, trying to tread the line between staying true to open source ideals and at the same time promoting our product to an audience that we want to help in a sympathetic and welcome way. I'll admit we haven't _always_ gotten this part right, but things move quickly, and "standards" of what is acceptable change.

I wonder sometimes whether a brief, easily ignored message that is seldom seen is a small, or large a price to pay for the value contained in a piece of software (am I a hypocrite for thinking that the Capistrano post-install message is just-about on the right side of the line, but the Java installer trying to install the Yahoo! toolbar should die in a fire?), that's largely a philosophical question which I'm afraid can't be well addressed in a GitHub issue.

> Additionally, making capistrano-harrow a direct and forced dependency of capistrano is equally over-the-top. And again, I don't think it'll get the desired effect. And again, imagine that installing rails came with a hard dependency on basecamp-api. I don't think anyone would find that in good taste either.

Actually this was done because we didn't want to invalidate people's "up to date" Capistrano installations by bundling the plugin in Capistrano itself. Making it a dependency with no fixed version should mean that people only need to see the post-install of `capistrano-harrow` once. I'd have liked to have solved this another way, but this was a limitation of Rubygems, which seems to have been largely misconstrued. Someone incorrectly posted to HackerNews "Capistrano adds gem dependency on third party paid service" which more than understandably caused a bit of outrage; the truth of the matter however was that we looked for a low-impact way to avoid bothering Capistrano users at large with the minutiae of our API integration. For what it's worth there were a number of tweaks to `capistrano-harrow` which would have mandated a bunch (as many as 10) of small tweaks to Capistrano itself had we bundled it. 

Interestingly, Mike Perham posted http://www.mikeperham.com/2016/02/09/kill-your-dependencies/ shortly before we launched this integration and the mood was still prevailing in the Ruby community when we shipped the version of Capistrano with this extra dependency, bad timing on our part, perhaps. Mike raises a good point though, fewer dependencies are usually better. No argument!

Optional gem dependencies would be a lovely feature of Rubygems, it's a feature we could have used here, and that we could have desperately needed in a [recent issue with Net::SSH and eliptical curve crypto][net-ssh-issue].

> I think it's counterproductive to attempt to force people to use something rather than to entice them to do so. Harrow looks like a lovely product. It's got the authority of capistrano maintainers behind it. There are better ways to draw that connection than by attempting to shove it down people's throat.

Already answered this above, we've had orders of magnitude more positive feedback about this integration than negative, but I'm well aware that there's a silent minority who might be festering in their own anger about feeling slighted without speaking up (in the age of transparency and easy contact to open source maintainers such as GitHub/etc there's no excuse for this, really) but we try incredibly hard to run an open and democratic project (2/3 of our core maintainers with commit and rubygems access don't work for Harrow, for example) that if anyone had a strong, strong disagreement they could easily have sent a PR to remove one line of code from Capistrano's gemspec with a small justification. It might not have been accepted, but the discussion has never really been had.

> I'd be happy to help brainstorm better ways.

I'd love to take you up on that! I'll feel better about doing so once we're a largely open source product. Whilst as a businessman and entrepreneur I'd love to get input from someone with your acumen, I'd feel much better about you using your time for the greater good of an open source product with a business attached.

> I hope we can find some better ways. Otherwise I don't see how Rails can continue to promote Capistrano in the default Gemfile or documentation. That would be a shame.

Truth be told, with the shifting tides in devops towards Docker, having rapidly passed and largely failed to notice Ansible, Salt and the cacophony of other similar things, I'd assumed that Capistrano still being in the Rails Gemfile was an oversight, rather than a note of continued conscious endorsement. That assumption not withstanding we're grateful of the continued support, and we have not forgotten our heritage.

In summary then if I may:

I see your point, I hope I've made justifications that help you better understand why certain decisions were made. I'm sorry that you feel some slight emotional distress, but I can understand your meta feeling of ownership and responsibility given that you fathered the entire community, and Capistrano is one of it's cornerstones. I'd be glad to engage with you to find alternative solutions in the short term, and to benefit from your experience and community reach in the medium term.

Thanks for taking the time to write harsh but clear words.

[net-ssh-issue]: https://github.com/net-ssh/net-ssh/issues/476

**Comment by dhh** [comment]
@leehambley, thanks for taking the time to go through the points. I'm a little unclear where this ultimately leaves us on the proposals. Are you open to nixing the install advertisement and making capistrano-harrow a separate gem, that someone would install on their own upon wishing to use Capistrano with Harrow?

If so, I'm happy to help with exposure through Twitter and I think we can do a profile on the Rails blog as well. I'd like to feature more companies that work with and have a stake in the Ruby and Rails ecosystem and should have our support.

If you need some time to think about this, that's fine too. If you could give me an estimate on how long you guys need, we can follow up from there.

**Comment by leehambley** [comment]
> I'm a little unclear where this ultimately leaves us on the proposals. Are you open to nixing the install advertisement and making capistrano-harrow a separate gem, that someone would install on their own upon wishing to use Capistrano with Harrow?

Definitely open to it. Ideally the tweet and any possible feature on the rails blog would ideally coincide with our open source announcement (from a karma, and business perspective). So far I know only that this will be "some time in February" and can't commit to anymore more concrete.

> If so, I'm happy to help with exposure through Twitter and I think we can do a profile on the Rails blog as well. I'd like to feature more companies that work with and have a stake in the Ruby and Rails ecosystem and should have our support.

That would be really cool, not just to be featured, but such a series in general. I've been fortunate to cross paths with the Passenger guys who have a niche and tricky business with a heavy reliance and investment on the Rails ecosystem, and Mike Perham of Sidekiq, I've always enjoyed insights into their inner workings.

Suggestion: 

1. Feel free to prepare a PR for Capistrano (or I will do it before the next release window), unfortunately you raise this issue ~3 days after our last release, next release is *latest* in six weeks according to our desired cadence. We'd release earlier if there was a noteworthy feature implemented, we probably wouldn't ship a release just for this, but the other guys may see this as a big enough change to warrant a deploy.
2. I'll follow up with you about your kind offers in due course, email or twitter when we've got something worth shouting about (estimate sometime during the next four weeks)

**Comment by dhh** [comment]
That's great to hear, @leehambley. What's the open source announcement you mention? 

Anyway, as long as we can arrive on an agreement to the two changes happening, I'm happy to start organizing the other elements we're talking about. Waiting 6 weeks for the trigger to be pulled on this wouldn't be a problem. But I'd want to arrive at a conclusion prior to that, just in case your "pushy investors" weren't on board, and we needed to make changes to the default Rails 5.1 Gemfile.

Completely agree about Passenger and Sidekiq. Those were exactly the two stories I was thinking about telling on the blog as well. I think those are good examples about optional extras to the Ruby and Rails ecosystem working as commercial offers.

Feel free to email me at david@basecamp.com for further coordination.

**Comment by leehambley** [comment]
> What's the open source announcement you mention?

At the risk of pre-mentioning it here, we're open-sourcing Harrow (more or less) in it's entirety. We've a secure enough market position, and firmly believe we, and everyone else will benefit from open sourcing the core of our service. (As well you know, most of the reasons people pay for hosted software hinge around ease-of-use and reliability engineering of the company providing the service) It's a step I'm very, very excited about!

**Comment by leehambley** [comment]
As it happens I'm on-board, I'm meeting the investor round on Friday this week, so let's agree on talking before Febuary ends at the *very* latest? I'll set a reminder.

**Comment by dhh** [comment]
Excellent. Best of luck and talk soon!

**Comment by leehambley** [comment]
Hi @dhh https://github.com/capistrano/capistrano/pull/1854 removes the gem dependency. Next release of Cap will be approx ~10th March 2017 given our six week cadence.

**Comment by leehambley** [comment]
Ohh, regarding the other points, I'll follow-up with you off-list as soon as I have time, and things are relevant, thanks again.

**Comment by dhh** [comment]
Great news, Lee! ❤️

**Comment by mattbrictson** [comment]
FYI, we [released capistrano 3.8.0 today](https://github.com/capistrano/capistrano/releases/tag/v3.8.0) with the harrow dependency [removed](https://github.com/capistrano/capistrano/pull/1854).

**Comment by dhh** [comment]
👏
